_posts/2019-07-23-micropython-pure-display-driver.md

```diff
@@ -21,22 +21,23 @@ So if you use ESP32 and you want to display a nice GUI on your ILI9341 display,
 
 LittlevGL project [already provides C implementation of that driver](https://github.com/littlevgl/esp32_ili9341), and there's even a [micropython API for it](https://github.com/littlevgl/lv_binding_micropython/blob/master/driver/esp32/modILI9341.c), but I wanted to create a **Pure Micropython implementation** of the same driver.
 
-This driver uses DMA and double buffering to acheive high performance.  
+This driver uses DMA and double buffering to achieve high performance.  
 At the bottom of this post you can find performance statistics and conclusions about using Micropython for a display driver.
 
 This display driver can run in two modes:
 - Pure Micropython mode - **all** the display driver logic is done in Micropython.
-- Hybrid mode - Setup and initialization is done in Micropython, but the critical path (flush and ISR functions) is implemented in C.
+- Hybrid mode - Setup and initialization are done in Micropython, but the critical path (flush and ISR functions) is implemented in C.
 
 
 # Why?
 
-Why would I re-write in micropython a module that is already written in C?  
+Why would I re-write a module in Micropython that is already written in C?  
 
-For two reasons:
+For three reasons:
 
  - It's a showcase for many interesting/advanced features related to Micropython Binding
  - I wanted to see how it would perform, compared to the C driver
+ - It makes possible to dynamically load drivers
```

Comment:
@kisvegabor This is possible with the existing C driver as well. The existing C driver has Micropython API so you can load it dynamically. So I don't think we should list this as a third reason.

Reply:
Maybe, it wasn't clear enough. I meant, for example, downloading drivers from a server and loading them run-time to support new displays. Like you download the driver for your video card on PC.

Reply:
I see.  
It's true, but it still requires a few non-obvious things on compile time:
- Configure `LV_COLOR_DEPTH=16` and `LV_COLOR_16_SWAP=1`
- ESP-IDF micropython module, including the [non-trivial C code to allow registering ISR](https://github.com/littlevgl/lv_binding_micropython/blob/806d33e074eabb5e38774ca247d5c6a2d2292eec/driver/esp32/espidf.c#L30-L118).  
- [Workaround](https://github.com/littlevgl/lv_micropython/commit/fe7db21f5e7410e6c3f9db4084fb2e21b0227ef1) on Micropython, to prevent [a crash](https://github.com/micropython/micropython/issues/4895#issuecomment-515637848).

If these are present on compile time, you can really load the driver on run time.